### ISSUE TITLE
Why is this module dependent on hiera?

### ISSUE THREAD

**Issue by seanscottking** [mayflower/puppet-php#118]
First of all: thank you for this module.

I'm surprised that this module is dependent on Hiera. Especially because this module has the "approved" badge on Puppet Forge. I would have expected Puppet Labs to have been a little more strict with it's review.

I'm sure that you already know that most people these days would recommend against doing Hiera calls within the module for various reasons and I don't want to beat that dead horse. Can you please explain what your reasoning is for doing it? Would you like some PRs to retire the dependency on Hiera? The good news is you already have a `params.pp`, which is usually where the function of Hiera calls should live.

Let us know your thoughts. I am a user of around ~30 component modules off of the Forge and this is the only one that does Hiera calls directly

**Comment by joth** [comment]
+1. I was puzzled by this too.

**Comment by globin** [comment]
**tl;dr**: we don't require hiera, we just work around [HI-118](https://tickets.puppetlabs.com/browse/HI-118)

There is a "bug/feature request" in hiera that breaks `deep_merge`ing in data binding even if hiera is told to use this feature. As we excessively use this feature in larger hierarchies, we work around this by explicitly merging the parameters by calling `hiera_hash` in these places. **BUT** we always also merge the parameter into that hash to, to always ensure that data_binding with a different terminus or explicitly passing parameters will work regardless of hiera being in use or not.

I personally think we should have caught all edge cases but if you find some unexpected behaviour or bugs please let us know, we are happy to fix things caused by our work around as long it doesn't break our merging. :wink: :beers:

@codec, @fadenb, @fpletz correct me if i got something wrong in the technical stuff.

**Comment by seanscottking** [comment]
Would you mind keeping this open, so we can get a few more opinions?

**Comment by fpletz** [comment]
We should in fact overhaul the documentation to reflect that hiera is not strictly needed anymore. Also, it would make sense to check if this is in fact true for at least all defined resources and classes that should be part of the public API. Some examples in the docs wouldn't hurt either.

**Comment by seanscottking** [comment]
OK @fpletz, I will test all public defined resource types and classes and report back here. I'll also submit a PR to the `README.md` with some examples.

**Comment by seanscottking** [comment]
My first attempt at PR to change code got closed without merging, so I am proposing a simpler change: just updating the `README.md` instead.

Even if/after you accept the README changes, I think we should still leave this issue open until we find out how to remove references of "hiera" within this module's code. I hope you agree.

**Comment by fpletz** [comment]
Why would you want to remove hiera from the module code completely? This is not the point I was making! It should just be possible to use all features and options of this module without hiera. If you think that is already the case this ticket can be closed.

Apart from that, your PR for updating the `README.md` is very welcome. Thanks!

**Comment by seanscottking** [comment]
@fpletz perhaps I should have linked this earlier. It is a few select links from a full write up by Puppet Labs' employee Gary Larizza. His blog is where I (and the others that have participated in here and on related PRs) are getting these ideas from.

In no particular order:
http://garylarizza.com/blog/2013/12/08/when-to-hiera/
http://garylarizza.com/blog/2014/10/24/puppet-workflows-4-using-hiera-in-anger/
http://garylarizza.com/blog/2014/02/17/puppet-workflow-part-1/
http://garylarizza.com/blog/2014/02/17/puppet-workflow-part-2/

**Comment by globin** [comment]
Please look at [HI-118](https://tickets.puppetlabs.com/browse/PUP-5395)

**Comment by globin** [comment]
Please look at [HI-118](https://tickets.puppetlabs.com/browse/PUP-5395)
